File: Leetcode2023/April2023/Java/NumberOfEnclaves.java
class Solution {
    public int numEnclaves(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        
        boolean[][] vis = new boolean[m][n];
        
        int[] dx = {0, 1, 0, -1};
        int[] dy = {1, 0, -1, 0};
        
        for (int i = 0; i < m; i++) {
            if (grid[i][0] == 1 && !vis[i][0]) {
                dfs(i, 0, grid, vis, dx, dy);
            }
            if (grid[i][n - 1] == 1 && !vis[i][n - 1]) {
                dfs(i, n - 1, grid, vis, dx, dy);
            }
        }
        
        for (int j = 0; j < n; j++) {
            if (grid[0][j] == 1 && !vis[0][j]) {
                dfs(0, j, grid, vis, dx, dy);
            }
            if (grid[m - 1][j] == 1 && !vis[m - 1][j]) {
                dfs(m - 1, j, grid, vis, dx, dy);
            }
        }
        
        int count = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1 && !vis[i][j]) {
                    count++;
                }
            }
        }
        
        return count;
    }
    
    private void dfs(int row, int col, int[][] grid, boolean[][] vis, int[] dx, int[] dy) {
        vis[row][col] = true;
        
        for (int i = 0; i < 4; i++) {
            int newRow = row + dx[i];
            int newCol = col + dy[i];
            if (isValid(newRow, newCol, grid, vis)) {
                dfs(newRow, newCol, grid, vis, dx, dy);
            }
        }
    }
    
    private boolean isValid(int row, int col, int[][] grid, boolean[][] vis) {
        int m = grid.length;
        int n = grid[0].length;
        return row >= 0 && row < m && col >= 0 && col < n && grid[row][col] == 1 && !vis[row][col];
    }
}

File: Leetcode2023/April2023/Python/NumberOfEnclaves.py
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:        
        m = len(grid)
        n = len(grid[0])

        vis = [[False for j in range(n)] for i in range(m)]

        dx = [0, 1, 0, -1]
        dy = [1, 0, -1, 0]

        for i in range(m):
            if grid[i][0] == 1 and not vis[i][0]:
                self.dfs(i, 0, grid, vis, dx, dy)

            if grid[i][n - 1] == 1 and not vis[i][n - 1]:
                self.dfs(i, n - 1, grid, vis, dx, dy)

        for j in range(n):
            if grid[0][j] == 1 and not vis[0][j]:
                self.dfs(0, j, grid, vis, dx, dy)

            if grid[m - 1][j] == 1 and not vis[m - 1][j]:
                self.dfs(m - 1, j, grid, vis, dx, dy)

        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1 and not vis[i][j]:
                    count += 1

        return count

    def dfs(self, row: int, col:int, grid: List[List[int]], vis: List[List[bool]], dx: List[int], dy: List[int]) -> None:

        vis[row][col] = True

        for i in range(4):
            newRow = row + dx[i]
            newCol = col + dy[i]

            if self.isValid(newRow, newCol, grid, vis):
                self.dfs(newRow, newCol, grid, vis, dx, dy)

    def isValid(self, row:int, col:int, grid: List[List[int]], vis: List[List[bool]]) -> bool:
        m = len(grid)
        n = len(grid[0])

        return row >= 0 and row < m and col >= 0 and col < n and grid[row][col] == 1 and not vis[row][col]

